src/Toolbox/Util/Graph/BFS.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE BangPatterns #-}
+
+-- | Breadth-First-Search (BFS)
+module Toolbox.Util.Graph.BFS
+  ( -- * internal state for BFS
+    BFSState (..),
+    BFSPath (..),
+    MultiBFSPath (..),
+
+    -- * single-seed use case
+    stepBFS,
+    runStagedBFS,
+
+    -- * multi-seed use case
+    stepMultiseedBFS,
+    runMultiseedStagedBFS,
+  )
+where
+
+import Control.Monad (join)
+import Control.Monad.Extra (loopM)
+-- import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Trans.State
+  ( StateT (..),
+    evalStateT,
+    get,
+    modify',
+  )
+import Data.Bifunctor (bimap)
+import Data.Either (partitionEithers)
+import qualified Data.Foldable as F
+import Data.IntMap (IntMap)
+import qualified Data.IntMap as IM
+import Data.IntSet (IntSet)
+import qualified Data.IntSet as IS
+import Data.Maybe (maybeToList)
+
+newtype BFSState = BFSState
+  { bfsVisited :: IntSet
+  }
+  deriving (Show)
+
+data BFSPath = BFSPath
+  { bfsSearchResult :: ![[Int]]
+  , bfsNextStage :: ![Int]
+  }
+  deriving (Show)
+
+stepBFS ::
+  forall m.
+  (Monad m) =>
+  IntMap [Int] ->
+  BFSPath ->
+  StateT BFSState m (Either BFSPath [[Int]])
```

Comment:
wow... what does the return type even mean? o.O

Reply:
what does it mean when you get a `Left path` out of this thing?

Reply:
This `step..` function is used for `Control.Monad.Extra.loopM` where `Either` is used to indicate whether the process is still in progress or done. 
`Left` is in progress (i.e. search is not over), and `Right` is done, so the search  result is returned. When done, the next stage information in `BFSPath` is not meaningful (it should be just empty list to finish), so only [[Int]] is passed. 
I admit that this needs some explanation as I described here now. I will add comments. Thank you!!

Reply:
added comment.